packages/langium/src/lsp/completion/completion-provider.ts

```diff
@@ -222,13 +225,29 @@ export class DefaultCompletionProvider implements CompletionProvider {
         const textDocument = document.textDocument;
         const text = textDocument.getText();
         const offset = textDocument.offsetAt(position);
-        const { nextTokenStart, nextTokenEnd, previousTokenStart, previousTokenEnd } = this.backtrackToAnyToken(text, offset);
         const partialContext = {
             document,
             textDocument,
             offset,
             position
         };
+        // Data type rules need special handling, as their tokens are irrelevant for completion purposes.
+        // If we encounter a data type rule at the current offset, we jump to the start of the data type rule.
+        const dataTypeRuleOffsets = this.findDataTypeRuleStart(cst, offset);
```

Comment:
What is happening here? When we have a FQN defined as:

```ts
FQN returns string: ID ('.' ID)*;
```

And then request a completion on `Example.For.Full<|>_Qualified.Name`, we get the entire range `<|Example.For.Full_Qualified.Name|>`... within are features like `ID` or?

Reply:
Yes, that correct. The `findDataTypeRuleStart` method searches the CST until it finds the top most node of the data type rule and returns the full range.